src/shell/xdg/popup.rs

```diff
@@ -0,0 +1,260 @@
+            + Dispatch<xdg_surface::XdgSurface, PopupData>
+            + Dispatch<xdg_popup::XdgPopup, PopupData>
+            + PopupHandler
+            + 'static,
+    {
+        let surface = compositor.create_surface(qh)?;
+        let popup = Self::from_surface(Some(parent), position, qh, surface, wm_base)?;
+        popup.wl_surface().commit();
+        Ok(popup)
+    }
+
+    /// Create a new popup from an existing surface.
+    ///
+    /// If you do not specify a parent surface, you must configure the parent using an alternate
+    /// function such as [`LayerSurface::get_popup`] prior to committing the surface.
+    ///
+    /// [`LayerSurface::get_popup`]: wayland_protocols_wlr::layer_shell::v1::client::zwlr_layer_surface_v1::ZwlrLayerSurfaceV1::get_popup
+    pub fn from_surface<D>(
+        parent: Option<&xdg_surface::XdgSurface>,
+        position: &xdg_positioner::XdgPositioner,
+        qh: &QueueHandle<D>,
+        surface: wl_surface::WlSurface,
+        wm_base: &XdgShellState,
+    ) -> Result<Popup, GlobalError>
+    where
+        D: Dispatch<xdg_surface::XdgSurface, PopupData>
+            + Dispatch<xdg_popup::XdgPopup, PopupData>
+            + 'static,
+    {
+        let arc = Arc::new(OnceCell::new());
+
+        let surface =
+            wm_base.create_xdg_surface(qh, surface, PopupData { inner: Arc::downgrade(&arc) })?;
+        let xdg_popup = surface.xdg_surface().get_popup(
+            parent,
+            position,
+            qh,
+            PopupData { inner: Arc::downgrade(&arc) },
+        )?;
+        arc.set(PopupInner {
+            surface,
+            xdg_popup,
+            pending_position: (AtomicI32::new(0), AtomicI32::new(0)),
+            pending_dimensions: (AtomicI32::new(-1), AtomicI32::new(-1)),
+            pending_token: AtomicU32::new(0),
+            configure_state: AtomicU32::new(0),
+        })
+        .unwrap();
+        Ok(Popup { inner: arc })
+    }
+
+    pub fn xdg_popup(&self) -> &xdg_popup::XdgPopup {
+        &self.inner.get().unwrap().xdg_popup
+    }
+
+    pub fn xdg_shell_surface(&self) -> &XdgShellSurface {
+        &self.inner.get().unwrap().surface
+    }
+
+    pub fn xdg_surface(&self) -> &xdg_surface::XdgSurface {
+        self.inner.get().unwrap().surface.xdg_surface()
+    }
+
+    pub fn wl_surface(&self) -> &wl_surface::WlSurface {
+        self.inner.get().unwrap().surface.wl_surface()
+    }
+
+    pub fn reposition(&self, position: &xdg_positioner::XdgPositioner, token: u32) {
+        self.xdg_popup().reposition(position, token);
+    }
+}
+
+impl Drop for PopupInner {
+    fn drop(&mut self) {
+        self.xdg_popup.destroy();
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct PopupConfigure {
+    /// (x,y) relative to parent surface window geometry
+    pub position: (i32, i32),
+    pub width: i32,
+    pub height: i32,
+    pub serial: u32,
+    /// Only valid if [Self::is_reposition]
+    pub token: u32,
+    state: u32,
```

Comment:
If this is internal data, this could probably just be an enum unless multiple states could be set at the same time.

Reply:
It's pulled from an `Atomic` and it can't be an enum when stored there.  I don't see much benefit from making it an enum just for half of the use, although it could have named constants.